### PR TITLE
Min at zero - design doc

### PR DESCRIPTION
**Pull request by mwielgus**
Ref: #43
cc: @MaciekPytel

### CODE REVIEW

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# Cluster Autoscaler - min at 0 
+### Design Document for Google Cloud Platform
+##### Author: mwielgus
+
+# Introduction
+
+One of the common requests for Cluster Autoscaler (for example:  [1], [2])  is the ability to scale some node groups to zero. This would definitely be a very useful feature but the implementation is somehow problematic in ScaleUP due to couple reasons:
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Either actually link examples or remove 'for example' part.

Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/somehow/somewhat

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+* [P1] There is no live example of what a new node would look like if the currently zero-sized node group was expanded. The node shape is defined as:
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Can we use some other letter for numbering problems (maybe I for 'issue')? I think most people interpret P1 as priority.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  * [P1A] Node capacity
+  * [P1B] Node allocatable
+  *  [P1C] Node labels
+* [P2] There is no live example of what manifest-run pods would be present on the new node.
+* [P3] There is no live example of what DaemonSets would be run on the new node.
+
+In general the above can be summarized as that the full definition of a new node needs to be somehow known before the node is actually created in order to decide whether the creation of a new node from a particular node group makes sense or not. Scale down has no issues with min@0.
+Design
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: /s/Design/# Design

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+Problems P1, P1A, P1B, P1C, P2, P3 needs to be solved. The primary focus is to create a solution for GCE/GKE but the proposed option should be generic enough to allow to expand this feature to other cloud providers if found needed and business-justified.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/needs/need

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  },
+```
+
+So all standard type sizes can be obtained by API. On GCP there is a possibility of a custom machine type. This custom machine type has the name formatted as:
+
+```
+custom-<cpu_count>-<memory_in_mb>
+```
+
+So it also quite easy to get all of capacity information from it.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/it/it is

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+### [P1B] - Node allocatable
+
+In GKE 1.5.6 allocatable for new nodes is equal to capacity. For simplicity we could assume that the new node will have the 90% (or -0.1cpu/-200mb) of capacity. Being wrong or underestimating here is not fatal, most users will probably be OK with this. Once some nodes are present we will have more precise estimates. The worst thing that can happen is that the scale up may not be triggered if the request is exactly at the node capacity - system pods.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: That is not true on non-GKE cluster on GKE. We could at least mention that fact and put some sort of TODO to revisit this.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+The amount is fixed at 100m and I guess it is unlikely to change so it can be probably hardcoded in CA as well.
+	
+### [P3] - There is no live example of what DaemonSets would be run on the new node.
+
+All daemon sets can be listed from apiserver and checked against the node with all the labels, capacities, allocatables and manifest-run pods obtained in previous steps. CA codebase already has the set of predicates imported so checking which pods should run on the node will be relatively easy.
+
+# Solution
+
+Given the all information above it should be relatively simple to write a module that given the access to GCP Api and Kubernetes API server. We will expand the NodeGroup interface (https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L40) with a method EstimateNodeShape, taking no parameters and returning NodeInfo (containing api.Node and all pods running by default on the node) or error if unable to do so.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Update the method name

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+On GCP the method will get the MIG template and perform analysis needed to estimate the node shape.
+
+The method will be called here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/utils.go#L132
+
+only if the current size of the node group is 0 or all of the nodes are unready/broken. Otherwise CA will try to estimate the shape of the node using live examples to avoid repeating any mis-estimation errors.
+
+The EstimateNodeShape will also be run on CA startup to ensure that CA is able to build an example for the node pool should the node group min size was set to 0.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: should the node group min size was set to 0 - that is not grammatical

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Done.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Done.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Done.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 6 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: added.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 7 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Done.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md (hunk 8 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: Done.

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+  * [1B] Node allocatable
+  *  [1C] Node labels
+* [2] There is no live example of what manifest-run pods would be present on the new node.
+* [3] There is no live example of what DaemonSets would be run on the new node.
+
+In general the above can be summarized as that the full definition of a new node needs to be somehow known before the node is actually created in order to decide whether the creation of a new node from a particular node group makes sense or not. Scale down has no issues with min@0.
+
+# Design
+
+Problems P1, P1A, P1B, P1C, P2, P3 needs to be solved. The primary focus is to create a solution for GCE/GKE but the proposed option should be generic enough to allow to expand this feature to other cloud providers.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/P1/1, etc

Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/primary focus/primary focus of this document

File: cluster-autoscaler/proposals/min_at_zero_gcp.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  },
+```
+
+So all standard type sizes can be obtained by API. On GCP there is a possibility of a custom machine type. This custom machine type has the name formatted as:
+
+```
+custom-<cpu_count>-<memory_in_mb>
+```
+
+So it ia also quite easy to get all of the capacity information from it.
```
Comment (MaciekPytel) [inline cluster-autoscaler/proposals/min_at_zero_gcp.md]: s/ia/is

**Pull request merged**